src/oge/data_cleaning.py

```diff
@@ -1784,43 +1784,190 @@ def remove_cems_with_zero_monthly_data(cems):
+    logger.info(
+        f"Memory usage after removing zero data: {post_memory_usage_gb:.2f} GB ({(post_memory_usage_gb / pre_memory_usage_gb * 100):.2f}% of original)"
+    )
     # drop the missing data flag column in-place
-    cems.drop(columns="missing_data_flag", inplace=True)
+    cems.drop(columns="zero_data_flag", inplace=True)
 
     return cems
 
 
+def complete_hourly_timeseries(
+    df: pd.DataFrame,
+    year: int,
+    group_cols: list[str] = [],
+    columns_to_fill_with_zero: list[str] = [],
+    columns_to_bffill: list[str] = [],
+) -> pd.DataFrame:
+    """
+    Completes a hourly timeseries for each plantgroup in a given dataframe.
+
+
+    This function will create a complete timeseries for each group in the dataframe,
+    assuming the data is supposed to represent a complete year. It will create a complete
+    timeseries for each group in UTC time.
+
+    Because we are repairing the "datetime_utc" column, but the complete set of utc
+    timestamps for a year depends on the local timezone of the underlying data, we
+    have to consider the timezone of each plant_id_eia in the data, and assign the
+    appropriate complete timeseries to each. This approach is more robust than just
+    creating a complete date_range between the min and max datetimes already in the
+    timeseries, in case that timeseries is missing timestamps at the beginning or end
+    of the year.
+
+    If "report_date" is passed in as one of the `group_cols`, the behavior of this
+    function is that it will only complete timeseries for "report_date"s that already
+    exist in `df`. For example, if `df` only contained data for January-June, if
+    "report_date" is included, then this function will only repair hourly timeseries for
+    January-June, but will not add timestamps for July-December. If that same df were
+    passed in without "report_date" in the `group_cols`, the function would also add
+    hourly timestamps for July-December. This functionality exists so that when we repair
+    `combined_cems_subplant_data` before combining it with the shaped eia data in step
+    18, we don't end up with overlapping timeseries.
+
+    Args:
+        df: dataframe to complete the timeseries for
+        year: year to create a complete timeseries for (should match the local year of the data)
+        group_cols: columns to group by (one column must be "plant_id_eia")
+        columns_to_fill_with_zero: a list of columns (generally containing numeric data)
+            that should be filled with zero values for missing timestamps that are filled
+        columns_to_bffill: a list of columns that should be filled based on the values in the previous and next rows within each group (bbffill refers to both bfill and ffill)
+            in the previous and next rows within each group (bbffill refers to both bfill
+            and ffill)
+    Returns:
+        dataframe with complete timeseries
+    """
```

Comment:
Missing typehints in docstring, e.g., `df (pd.DataFrame):` or `pd.DataFrame: dataframe with complete timeseries` for the returned value